iwyu.cc

```diff
@@ -2585,11 +2603,17 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
 
   const UsingDecl* GetUsingDeclarationOf(const NamedDecl* decl,
                                          const DeclContext* using_context) {
-    // We look through all the using-decls of the given decl.  We
-    // limit them to ones that are visible from the decl-context we're
-    // currently in (that is, what namespaces we're in).  Of those, we
-    // pick the one that's in the same file as decl, if possible,
-    // otherwise we pick one arbitrarily.
+    // First, if we have a UsingShadowDecl, then we don't need to do anything
+    // because we can just directly return the using decl from that.
+    if (const UsingShadowDecl* shadow = DynCastFrom(decl))
+      return shadow->getUsingDecl();
+
+    // But, if we don't have a UsingShadowDecl, then we need to look through
+    // all the using-decls of the given decl.  We limit them to ones that are
+    // visible from the decl-context we're currently in (that is, what
+    // namespaces we're in), via the check through 'Encloses'. Of those, we
```

Comment:
I'm unable to break it. Everything I do that seems should require equials, passes with or without it, so I think we should leave it as is for now.

Interestingly, along the way, the only case that I could find that engages this code path is Variable declarations.  Everything else ends up passing though VisitDeclRefExpr, which catches the ShadowDecls.  It would be interesting to know *why* VarDecls aren't also caught by DeclRefExpr because if it's a bug in Clang, fixing it might remove the need for this map entirely.